### PR TITLE
xenserver_guest_info: use fallback chain for VDI format detection

### LINKED ISSUE #11998: Add support for disks VDI UUID and VDI type VHD/QCOW2 in community.general.xenserver_guest_info module
### Summary

We are doing a lot of VM migrations from and to XenServer / XCP-ng using ansible playbooks and it would be very useful for us to know VDI UUID for each disk returned via xenserver_guest_info module. Currently it returns
```
"disks": [
    {
        "name": "devel-nx-20_10-disk0",
        "name_desc": "",
        "os_device": "",
        "size": 21474836480,
        "sr": "r-data2",
        "sr_uuid": "074dc759-d250-a9c9-2abb-177408bf7028",
        "vbd_userdevice": "0"
      }
],
```
but it would be nice to have
```
"disks": [
    {
        "name": "devel-nx-20_10-diski0",
        "name_desc": "",
        "os_device": "",
        "size": 21474836480,
        "sr": "r-data2",
        "sr_uuid": "074dc759-d250-a9c9-2abb-177408bf7028",
        "uuid": "08baf46c-306d-4aa8-aee0-20753abfcc54",
        "vbd_userdevice": "0"
      }
],
```
Also there is an upcoming support for QCOW2 disk image format in XCP-ng 8.3, maybe it is possible to add some field to display either VDI is in VHD or QCOW2 format.

### Issue Type

Feature Idea

### Component Name

plugins/module_utils/_xenserver.py

### Additional Information

```
--- a/plugins/module_utils/xenserver.py      2026-04-29 19:19:33.322307933 +0300
+++ b/plugins/module_utils/xenserver.py      2026-05-05 19:52:58.626270393 +0300
@@ -445,6 +445,7 @@
             vm_disk_sr_params = xapi_session.xenapi.SR.get_record(vm_vbd_params["VDI"]["SR"])

             vm_disk_params = {
+                "uuid": vm_vbd_params["VDI"]["uuid"],
                 "size": int(vm_vbd_params["VDI"]["virtual_size"]),
                 "name": vm_vbd_params["VDI"]["name_label"],
                 "name_desc": vm_vbd_params["VDI"]["name_description"],
```

### Code of Conduct

- [x] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by Shreyashxredhat**
##### SUMMARY

Follow-up to #12119 based on feedback from @bvitnik (https://xcp-ng.org/forum/topic/12259).

This PR improves VDI format detection in `xenserver_guest_info` by using the following fallback chain when reading `sm_config`:

`image-format` → `vdi_type` → `type` → `raw`

Additional changes:
- Remove the unnecessary `.get()` default for `sm_config`, since it is a mandatory parameter and always present.
- Update RETURN documentation examples in `xenserver_guest` and `xenserver_guest_powerstate` to include the `uuid` and `vdi_type` fields.

Fixes #11998

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

xenserver_guest_info
xenserver_guest
xenserver_guest_powerstate

##### ADDITIONAL INFORMATION

This is a follow-up cleanup and compatibility improvement based on testing against both XenServer and XCP-ng environments. The updated fallback logic ensures the correct VDI format is reported across different storage backends and platform versions.

### CODE REVIEW

**Comment by Shreyashxredhat** [comment]
Hey @bvitnik, Please take a look.

File: plugins/module_utils/_xenserver.py [hunk 1]
```diff
@@ -456,7 +456,13 @@ def gather_vm_facts(module: AnsibleModule, vm_params):
                 "os_device": vm_vbd_params["device"],
                 "uuid": vm_vbd_params["VDI"]["uuid"],
                 "vbd_userdevice": vm_vbd_params["userdevice"],
-                "vdi_type": vm_vbd_params["VDI"].get("sm_config", {}).get("vdi_type", ""),
+                "vdi_type": vm_vbd_params["VDI"]["sm_config"].get(
+                    "image-format",
+                    vm_vbd_params["VDI"]["sm_config"].get(
+                        "vdi_type",
+                        vm_vbd_params["VDI"]["sm_config"].get("type", "raw"),
```
Comment (russoz) [inline plugins/module_utils/_xenserver.py:463]: If this is for an Ansible parameter (any param/sub-param of the module itself), keep in mind that Ansible **always** create entries in the dictionary for all params, even when they have not been passed.

If that's the case here, it means that this `.get("type", "raw")` will **never** return "raw" (unless of course, the module user themself passes that value).

**Review by russoz — COMMENTED** [review]
Hi @Shreyashxredhat 

A comment there.

File: plugins/module_utils/_xenserver.py (hunk 1 above)
Comment (bvitnik) [reply plugins/module_utils/_xenserver.py:463]: No. These are not module params. These are XenAPI object related and are incidentally also called params.

Additionally, only the `sm_config` param, being the top level one, is documented and defined in the API documentation. By nature (or API design) it is of mapping type (string to string) which in XenAPI terms mean that keys and values inside `sm_config` are freely chosen. You can basically write anything there but only some keys and values have any real meaning or significance, either on the server or on the client side. API returning default values for any of these sub params is thus impossible.

**Comment by bvitnik** [comment]
@Shreyashxredhat thanks for making the PR. I'm thinking of two more things and any additional input is appreciated.

Fist, we could set the `vdi_type` key in `ansible-test-vm-3-params.json‎` fixture (for unit tests) to some improper value. That way we can test, at least partially, the chain resolution. `image-format` should take precedence in this case and proper value should be returned regardless of `vdi_type` having improper value. Yeah, for full coverage we would need more fixtures but this is good enough for me.

Second, I'm thinking if we should add some note to all three modules mentioning that `vdi_type` is heuristically determined and could be wrong. On the other hand, if we don't add any note, unsuspecting users could identify and report to us any edge cases where our fallback chain gives wrong result. Then, we can fix it. With note being there, users will just assume it is OK if it does not work correctly.

**Comment by Shreyashxredhat** [comment]
hey @bvitnik , i can do the test fixture change. For the documentation question, i would lean toward not adding a note for now, because i think without the note, users are more likely to report edge cases where the chain gives wrong results, which helps improve it.

File: plugins/module_utils/_xenserver.py (hunk 1 above)
Comment (russoz) [reply plugins/module_utils/_xenserver.py:463]: Thanks for clarifying! :-)

**Comment by bvitnik** [comment]
Great @Shreyashxredhat . Then we have an agreement :smile: . Code changes look fine to me and we will skip the note.

@russoz we can ship this :+1: .

**Comment by russoz** [comment]
@Shreyashxredhat thans for your contribution!
@bvitnik @felixfontein thanks for reviewing!
